drv/lpc55-sprot-server/src/main.rs

```diff
@@ -149,63 +154,47 @@ enum IoError {
 fn main() -> ! {
     let mut io = configure_spi();
 
-    let mut rx_buf = [0u8; BUF_SIZE];
-    let mut tx_buf = [0u8; BUF_SIZE];
+    let mut rx_buf = [0u8; MAX_REQUEST_SIZE];
+    let mut tx_buf = [0u8; MAX_RESPONSE_SIZE];
 
     let mut handler = Handler::new();
-    let mut signal_reply = false;
+
+    // Prime our write fifo, so we clock out zero bytes on the next receive
+    io.spi.drain_tx();
+    while io.spi.can_tx() {
+        io.spi.send_u8(0);
+    }
 
     loop {
-        // Every time through the io loop we receive a fresh request
-        let mut rx_msg = RxMsg::new(&mut rx_buf[..]);
-
-        let reply = match io.next(signal_reply, &tx_buf[..], &mut rx_msg) {
-            Ok(()) => {
-                // Put a busy byte in the tx_fifo before we handle the result,
-                // which may take some time.
-                io.mark_busy();
-                // Clear the buffer in preparation for a reply
-                let tx_msg = TxMsg::new(&mut tx_buf[..]);
-                handler.handle(rx_msg, tx_msg, &mut io.stats)
+        let rsp_len = match io.wait_for_request(&mut rx_buf) {
+            Ok(rx_len) => {
+                handler.handle(&rx_buf[..rx_len], &mut tx_buf, &mut io.stats)
+            }
+            Err(IoError::Flush) => {
+                ringbuf_entry!(Trace::Flush);
+                continue;
             }
-            Err(IoError::Flush) => None,
             Err(IoError::Flow) => {
-                io.mark_busy();
-                let tx_msg = TxMsg::new(&mut tx_buf[..]);
-                Some(handler.flow_error(tx_msg))
+                ringbuf_entry!(Trace::FlowError);
+                handler.flow_error(&mut tx_buf)
             }
         };
 
-        if reply.is_none() {
-            // There's no reply to send, so we should clear the Tx buffer.
-            tx_buf.fill(0);
-            signal_reply = false;
-        } else {
-            signal_reply = true;
-        }
+        ringbuf_entry!(Trace::Stats(io.stats));
+        io.reply(&tx_buf[..rsp_len]);
     }
 }
 
 impl Io {
-    /// Put a Protocol::Busy value in FIFOWR so that SP/logic analyzer knows
-    /// we're away.
-    pub fn mark_busy(&mut self) {
-        self.spi.drain_tx();
-        self.spi.send_u8(Protocol::Busy as u8);
-    }
-
-    pub fn next<'a>(
-        &mut self,
-        signal_reply: bool,
-        tx_buf: &[u8],
-        rx_msg: &mut RxMsg<'a>,
-    ) -> Result<(), IoError> {
-        let mut tx_iter = tx_buf.iter();
-        self.prime_tx_fifo(&mut tx_iter);
-        let result = loop {
+    // Wait for chip select to be asserted
+    // Assert ROT_IRQ if this is a reply
+    fn wait_for_csn_asserted(&mut self, is_reply: bool) {
+        let mut rot_irq_asserted = false;
```

Comment:
How do you know ROT_IRQ is not asserted at this point?

Reply:
To be clear, this was a bug I introduced in this refactoring, not an existing bug :)